### ISSUE TITLE
Reachability helper feature hangs tests in some scenario

### ISSUE THREAD

**Issue by meziantou** [coverlet-coverage/coverlet#1044]
After updating to Coverlet 3.0.0, one of the test projects doesn't run anymore. The process is stuck at:

````
Test run for C:\Users\mezia\source\repos\Meziantou.Framework\tests\Meziantou.Framework.SingleInstance.Tests\bin\Debug\net461\Meziantou.Framework.SingleInstance.Tests.dll (.NETFramework,Version=v4.6.1)
Microsoft (R) Test Execution Command Line Tool Version 16.8.3
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
````

The CPU usage is 25% for the .NET process (I have a 4 cores CPU).

I've generated a dump file, and it seems to be stuck somewhere in `Mono.Cecil`:

````
System.Private.CoreLib.dll!System.SpanHelpers.SequenceEqual(ref byte first, ref byte second, nuint length) Line 1561
	at /_/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs(1561)
Mono.Cecil.dll!Mono.Cecil.MetadataResolver.GetType(Mono.Cecil.ModuleDefinition module, Mono.Cecil.TypeReference reference)
Mono.Cecil.dll!Mono.Cecil.MetadataResolver.Resolve(Mono.Cecil.MethodReference method)
Mono.Cecil.dll!Mono.Cecil.ModuleDefinition.Resolve(Mono.Cecil.MethodReference method)
Mono.Cecil.dll!Mono.Cecil.MethodReference.Resolve()
coverlet.core.dll!Coverlet.Core.Instrumentation.Reachability.ReachabilityHelper.CreateForModule(Mono.Cecil.ModuleDefinition module, string[] doesNotReturnAttributes, Coverlet.Core.Abstractions.ILogger logger) Line 299
	at /_/src/coverlet.core/Instrumentation/ReachabilityHelper.cs(299)
coverlet.core.dll!Coverlet.Core.Instrumentation.Instrumenter.CreateReachabilityHelper() Line 210
	at /_/src/coverlet.core/Instrumentation/Instrumenter.cs(210)
coverlet.core.dll!Coverlet.Core.Instrumentation.Instrumenter.InstrumentModule() Line 219
	at /_/src/coverlet.core/Instrumentation/Instrumenter.cs(219)
coverlet.core.dll!Coverlet.Core.Instrumentation.Instrumenter.Instrument() Line 159
	at /_/src/coverlet.core/Instrumentation/Instrumenter.cs(159)
coverlet.core.dll!Coverlet.Core.Coverage.PrepareModules() Line 146
	at /_/src/coverlet.core/Coverage.cs(146)
coverlet.collector.dll!Coverlet.Collector.DataCollection.CoverageWrapper.PrepareModules(Coverlet.Core.Coverage coverage) Line 63
	at /_/src/coverlet.collector/DataCollection/CoverageWrapper.cs(63)
coverlet.collector.dll!Coverlet.Collector.DataCollection.CoverageManager.InstrumentModules() Line 65
	at /_/src/coverlet.collector/DataCollection/CoverageManager.cs(65)
coverlet.collector.dll!Coverlet.Collector.DataCollection.CoverletCoverageCollector.OnSessionStart(object sender, Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection.SessionStartEventArgs sessionStartEventArgs) Line 140
	at /_/src/coverlet.collector/DataCollection/CoverletCoverageCollector.cs(140)
[Native to Managed Transition]
[Managed to Native Transition]
System.Private.CoreLib.dll!System.Delegate.DynamicInvokeImpl(object[] args) Line 89
	at /_/src/coreclr/src/System.Private.CoreLib/src/System/Delegate.CoreCLR.cs(89)
System.Private.CoreLib.dll!System.Delegate.DynamicInvoke(object[] args) Line 64
	at /_/src/libraries/System.Private.CoreLib/src/System/Delegate.cs(64)
Microsoft.TestPlatform.CoreUtilities.dll!Microsoft.VisualStudio.TestPlatform.Utilities.MulticastDelegateUtilities.SafeInvoke(System.Delegate delegates, object sender, System.EventArgs args, string traceDisplayName)
Microsoft.VisualStudio.TestPlatform.Common.dll!Microsoft.VisualStudio.TestPlatform.Common.DataCollector.TestPlatformDataCollectionEvents.OnSessionStart(Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection.DataCollectionEventArgs e)
Microsoft.VisualStudio.TestPlatform.Common.dll!Microsoft.VisualStudio.TestPlatform.Common.DataCollector.TestPlatformDataCollectionEvents.RaiseEvent(Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection.DataCollectionEventArgs e)
Microsoft.VisualStudio.TestPlatform.Common.dll!Microsoft.VisualStudio.TestPlatform.Common.DataCollector.DataCollectionManager.SendEvent(Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection.DataCollectionEventArgs args)
Microsoft.VisualStudio.TestPlatform.Common.dll!Microsoft.VisualStudio.TestPlatform.Common.DataCollector.DataCollectionManager.SessionStarted(Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection.SessionStartEventArgs sessionStartEventArgs)
Microsoft.TestPlatform.CommunicationUtilities.dll!Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.DataCollection.DataCollectionRequestHandler.HandleBeforeTestRunStart(Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.Message message)
Microsoft.TestPlatform.CommunicationUtilities.dll!Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.DataCollection.DataCollectionRequestHandler.ProcessRequests()
datacollector.dll!Microsoft.VisualStudio.TestPlatform.DataCollector.DataCollectorMain.StartProcessing()
datacollector.dll!Microsoft.VisualStudio.TestPlatform.DataCollector.DataCollectorMain.Run(string[] args)
datacollector.dll!Microsoft.VisualStudio.TestPlatform.DataCollector.Program.Main(string[] args)
````

Environment:
- Windows 10
- .NET SDK 5.0.102

Repro steps:
- git clone https://github.com/meziantou/Meziantou.Framework.git
- cd Meziantou.Framework
- git checkout f96e0db0ab979a7d8344ec92391a1136b19a14fe
- dotnet build Meziantou.Framework.sln --configuration Debug /bl
- dotnet test .\tests\Meziantou.Framework.SingleInstance.Tests\Meziantou.Framework.SingleInstance.Tests.csproj  --configuration Debug --no-build --logger trx --collect:"XPlat Code Coverage" -f net461

=> This is the only project of the solution that hangs, and it works with other target frameworks (`-f net5.0`)
=> It works with coverlet 1.3.0
=> It works without `--collect:"XPlat Code Coverage"`

**Comment by MarcoRossignoli** [comment]
Thanks for reporting this, we moved collector to netstandard support netfx, that could be the reason. I had similar issue in past with loop resolution inside cecil related to different platform type forwarding.

**Comment by danielp37** [comment]
I'm seeing the same thing in a project that I'm working on.  With my projects, I've narrowed it down to a project that works with objects from System.Data (like DataTable or SqlConnection) AND the project also references one of these 3 nuget packages:

<PackageReference Include="Microsoft.TeamFoundationServer.ExtendedClient" Version="16.153.0" /> 
<PackageReference Include="Microsoft.VisualStudio.Services.Client" Version="16.153.0" />
 <PackageReference Include="Microsoft.VisualStudio.Services.InteractiveClient" Version="16.153.0" />

And these references need to be in a project that the unit test project references.  If the System.Data references are in the test project, then it works fine.  It has to be in a separate project.  And if I don't reference any of these three projects it works fine and if I reference the projects but don't access system.data, it works fine.  

I am targeting net461, am using SDK-style projects and running the tests with dotnet test targeting the .csproj.  And just like above, it works with coverlet 1.3.0 and works without --collect:"XPlat Code Coverage".

**Comment by bthharper** [comment]
Am using `nunit3-console` with the `coverlet` console command line rather than `dotnet test` but get the same result.  Have reverted back to 1.7.2 for now.

**Comment by MarcoRossignoli** [comment]
Can you all try with this packages?

[Packages.zip](https://github.com/coverlet-coverage/coverlet/files/5826139/Packages.zip)

**Comment by meziantou** [comment]
@MarcoRossignoli The CI completes with the preview package :)
https://github.com/meziantou/Meziantou.Framework/runs/1717298953

Thanks for the fix! I'll wait for the official release to update coverlet in my project

**Comment by MarcoRossignoli** [comment]
Unfortunately I fixed your issue but now I got a lot of noise from reachability helper, I need more time 😢 

https://dev.azure.com/tonerdo/coverlet/_build/results?buildId=1324&view=logs&j=f0f0520d-467b-5d1d-a95d-4e3a2581fe8a&t=67725203-aed8-540e-d7ff-b357d1276ce7

**Comment by bthharper** [comment]
I've tried the updated coverlet, but it still hangs for me,

```powershell
 	[External Code]	
 	[Managed to Native Transition]	
 	System.Private.CoreLib.dll!System.Collections.Generic.Dictionary<Mono.Cecil.Metadata.Row<string, string>, Mono.Cecil.TypeDefinition>.TryGetValue(Mono.Cecil.Metadata.Row<string, string> key, out Mono.Cecil.TypeDefinition value)	Unknown
 	Mono.Cecil.dll!Mono.Cecil.TypeDefinitionCollection.GetType(string namespace, string name)	Unknown
 	Mono.Cecil.dll!Mono.Cecil.MetadataResolver.GetType(Mono.Cecil.ModuleDefinition module, Mono.Cecil.TypeReference reference)	Unknown
 	Mono.Cecil.dll!Mono.Cecil.MetadataResolver.Resolve(Mono.Cecil.MethodReference method)	Unknown
 	Mono.Cecil.dll!Mono.Cecil.ModuleDefinition.Resolve(Mono.Cecil.MethodReference method)	Unknown
 	Mono.Cecil.dll!Mono.Cecil.MethodReference.Resolve()	Unknown
>	coverlet.core.dll!Coverlet.Core.Instrumentation.Reachability.ReachabilityHelper.CreateForModule(Mono.Cecil.ModuleDefinition module, string[] doesNotReturnAttributes, Coverlet.Core.Abstractions.ILogger logger) Line 299	C#
 	coverlet.core.dll!Coverlet.Core.Instrumentation.Instrumenter.CreateReachabilityHelper() Line 209	C#
 	coverlet.core.dll!Coverlet.Core.Instrumentation.Instrumenter.InstrumentModule() Line 218	C#
 	coverlet.core.dll!Coverlet.Core.Instrumentation.Instrumenter.Instrument() Line 159	C#
 	coverlet.core.dll!Coverlet.Core.Coverage.PrepareModules() Line 146	C#
 	coverlet.console.dll!Coverlet.Console.Program.Main.AnonymousMethod__1() Line 106	C#
 	McMaster.Extensions.CommandLineUtils.dll!McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(string[] args)	Unknown
 	coverlet.console.dll!Coverlet.Console.Program.Main(string[] args) Line 261	C#
 	[Native to Managed Transition]	
 	[External Code]	
```

**Comment by MarcoRossignoli** [comment]
I think for the moment I'll go for disable ["Methods that do not return"](https://github.com/coverlet-coverage/coverlet/blob/master/Documentation/MSBuildIntegration.md#methods-that-do-not-return) by default.
I need more time to understand what's happening, the type resolution part is a bit tricky and at the moment there is no workaround to unblock.
cc: @kevin-montrose

**Comment by zplan** [comment]
> Unfortunately I fixed your issue but now I got a lot of noise from reachability helper, I need more time 😢
> 
> https://dev.azure.com/tonerdo/coverlet/_build/results?buildId=1324&view=logs&j=f0f0520d-467b-5d1d-a95d-4e3a2581fe8a&t=67725203-aed8-540e-d7ff-b357d1276ce7
> 
> A lot of warning `[coverlet] Unable to resolve method reference...`

Confirm.
I see a lot of these "Unable to resolve..." warnings comming from coverlet.msbuild.targets(39,5)
Installed version: coverlet.msbuild 3.0.1

**Comment by MarcoRossignoli** [comment]
thx for test @zplan asap I'll attach a new test lib

**Comment by kevin-montrose** [comment]
Looks like it's getting stuck trying to resolve a call to `System.Security.Principal.WindowsIdentity System.Security.Principal.WindowsIdentity::GetCurrent()` which then gets stuck in a cycle of looking up `System.Security.Principal.Windows, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a` and `mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089` in `NetstandardAwareAssemblyResolver`.

Very weird.

`dotnet test` (and thus the coverlet code) are actually running under .NET 5 in this scenario, so my suspicion is we're looking in the wrong directories to resolve references.

Dropping some breakpoints around, I see 
![image](https://user-images.githubusercontent.com/1219467/105106598-34072b00-5a84-11eb-96df-80b3dd27dd34.png)

Which smells really wrong.

I'm not super clear on how `NetstandardAwareAssemblyResolver` is supposed to behave, but the reachability stuff merely copied the use in `Instrumenter.InstrumentModule()` so it seems likely a bigger problem than just a bug in reachability.

I'll keep pondering on this to see if I can't figure out what's going on with more precision...

**Comment by kevin-montrose** [comment]
Yeah, it looks like if I make sure the correct .net folder is being search the symbol lookup succeeds:
![image](https://user-images.githubusercontent.com/1219467/105108576-8e09ef80-5a88-11eb-8e5c-2f5e1405f6e9.png)

So we'll need some way to determine that from just `_module`.  I'll keep poking, though I'm running out of personal-hacking-time tonight...

**Comment by MarcoRossignoli** [comment]
> so it seems likely a bigger problem than just a bug in reachability

Yep it's not a bug in reachability for sure I found your same issue (loop on resolution). It's for sure related to resolution and seem only for net461.

 > NetstandardAwareAssemblyResolver

Is used for similar reason, as you can see from the comments above that class it loads correct netstandard assembly to allow correct forward type resolution between different runtimes.

**Comment by kevin-montrose** [comment]
Writing down my rough understanding, and the probable fix here.

 1. The loop happens because we're looking at the .NET version that's running the `dotnet test` command whenever we can't find an assembly in the test-containing-module's directory.
   a. So, in theory any symbol that's moved between assemblies between 4.6.1 and .NET 5 would trigger this.  This is vanishingly few symbols, but `System.Security.Principal.WindowsIdentity System.Security.Principal.WindowsIdentity` is one (in mscorlib.dll in 4.6.1, and System.Security.Principal.Windows.dll in 5)
  2. We end up looking at the running .NET (which is 5 now) because `NetstandardAwareAssemblyResolver` is falling back on `DefaultAssemblyResolver` (which itself falls back to [`BaseAssemblyResolver`](https://github.com/jbevain/cecil/blob/1b1b58a5da2a261d003f3f2647f7219b5344f20d/Mono.Cecil/BaseAssemblyResolver.cs#L137)).  That Cecil code consults trusted assemblies (as defined by an environment variable), not anything particular to the test DLL.
    a. The loop is kind of non-obvious, but is coming from loading each trusted assembly as part of [`SearchTrustedPlatformAssemblies`](https://github.com/jbevain/cecil/blob/1b1b58a5da2a261d003f3f2647f7219b5344f20d/Mono.Cecil/BaseAssemblyResolver.cs#L176) which ends up using the current `IAssemblyResolver`

I tried some hacks to detect this sort of looping, but I couldn't get anything reasonable to work.  Which makes me think the only fix is detecting the appropriate .NET folder to reference (if we testing framework) and just add it to the search directories.

I'mma give it a go for the rest of tonight, my recollection on determining .NET Framework install directories is that it... sucks.

**Comment by kevin-montrose** [comment]
Oy, something like this...

```csharp
private void AddModuleSearchDirectories(BaseAssemblyResolver resolver)
{
    resolver.AddSearchDirectory(Path.GetDirectoryName(_module));

    // if module is linked against a FRAMEWORK version we have to go and add it's location to search directories as well
    // we don't need to do anything with .net core or .net 5.0+ as we're running as one of them, and everything should be
    // forward compatible

    using (var stream = _fileSystem.NewFileStream(_module, FileMode.Open, FileAccess.Read))
    using (var module = ModuleDefinition.ReadModule(stream))
    {
        var version = module.Runtime;
        if (version != TargetRuntime.Net_4_0)
        {
            // we are in a weeeeeeird place, bail
            return;
        }

        var attrs = module.Assembly.CustomAttributes;
        var targetVersionAttr = attrs.SingleOrDefault(x => x.AttributeType.Namespace == "System.Runtime.Versioning" && x.AttributeType.Name == "TargetFrameworkAttribute");

        if (targetVersionAttr == null)
        {
            // this should be present for all 4.0+ and Core and Standard installs... so, if not just give up
            return;
        }

        var targetVersion = (string)targetVersionAttr.ConstructorArguments[0].Value;
        if (targetVersion.StartsWith(".NETFramework,"))
        {
            // every single version of .NET Framework 4.x stores it's info in the same place
            // ... the registry

            using (var key = Microsoft.Win32.Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Client", writable: false))
            {
                var frameworkPath = key?.GetValue("InstallPath") as string;
                if (frameworkPath == null)
                {
                    // whelp, .net isn't installed?
                    return;
                }

                resolver.AddSearchDirectory(frameworkPath);
            }
        }
        else
        {
            // if we start with something else, we're in netstandard, core, or .net 5+ land... nothing new to add
            return;
        }
    }
}
```

replacing the two existing calls to `AddSearchDirectory` seems to work.  I'm not in a great spot to run the full test suite, but the net461 and net5.0 cases in this issue do seem to work.

Frankly, it stinks (and requires taking a dependency [Microsoft.Windows.Registry](https://www.nuget.org/packages/Microsoft.Win32.Registry/)) but there just isn't a [great way to find .NET Framework install directories](https://docs.microsoft.com/en-us/troubleshoot/dotnet/framework/determine-dotnet-versions-service-pack-levels).

I don't know if more logging is desired in the weird-bail cases, and testing those cases will be painful (can `dotnet test` even run pre-4.0 assemblies?  I've never been in a position to try.).

Because Cecil falls back on consulting the current processes trusted assemblies, if we're looking at a netstandard/core/5+ `_module`  things just work without other changes.  This is subtle (and implicit) behavior, so I don't feel amazing about it.

**Comment by MarcoRossignoli** [comment]
@zplan @meziantou can you try with this preview version? 

[Preview.zip](https://github.com/coverlet-coverage/coverlet/files/5860348/Preview.zip)

Moved reachability helper opt-in until we properly fix the issue to avoid to block corner cases.

**Comment by meziantou** [comment]
@MarcoRossignoli The tests complete using the 3.0.2-preview.3 version :)

**Comment by MarcoRossignoli** [comment]
Just released new version https://github.com/coverlet-coverage/coverlet/releases/tag/v5.6.2

**Comment by meziantou** [comment]
I've merged the latest version and it seems ok on my side 😊

Thanks for the investigation and the quick fix!

**Comment by zplan** [comment]
> @zplan @meziantou can you try with this preview version?
> 
> [Preview.zip](https://github.com/coverlet-coverage/coverlet/files/5860348/Preview.zip)
> 
> Moved reachability helper opt-in until we properly fix the issue to avoid to block corner cases.

@MarcoRossignoli 
Sorry, had no time.

Thanks for fixing, I will check the new version!

**Comment by RomanBadiornyi** [comment]
Hi @MarcoRossignoli, I'm facing same issue. While digging I found out that process gets frozen at 
https://github.com/coverlet-coverage/coverlet/blob/3ae19761876f2ac809c227ca3563495e28993b59/src/coverlet.core/Instrumentation/Instrumenter.cs#L306

After I have injected "C:\Windows\Microsoft.NET\Framework64\v4.0.30319" into resolver 
`resolver.AddSearchDirectory("C:\Windows\Microsoft.NET\Framework64\v4.0.30319");`
it worked and write method completed within seconds. 

As I understand, we didn't took @kevin-montrose code due to additional dependency which isn't net standard of course. 
Do we have any other workaround how framework path can be injected?
Could we maybe consider defining it in the runSettings file? 

We are blocked right now due to this issue, I really appreciate any help here.

**Comment by MarcoRossignoli** [comment]
@RomanBadiornyi with last version the reachability helper is disabled by default so you shouldn't be stuck here. Try last version and let me know.

**Comment by RomanBadiornyi** [comment]
@MarcoRossignoli, I'm running coverlet.collecter 3.0.3. When you say latest, you mean some pre-release version?

**Comment by MarcoRossignoli** [comment]
@RomanBadiornyi  this PR https://github.com/coverlet-coverage/coverlet/pull/1061 removed temporary the reachability helper(if you're not using `DoesNotReturnAttribute` arg), so I think that your problem is not related to this issue if you're on last version.
As you can see from https://github.com/coverlet-coverage/coverlet/issues/1044#issuecomment-766392552 after patch should work.

Can you open a new issue and maybe try a repro?Maybe we can try to add that folder by default for .NET Framework app using @kevin-montrose's idea.

**Comment by RomanBadiornyi** [comment]
@MarcoRossignoli, so far I have found out that it 's keep switching between resolving path of 
System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
and
System.IO.Compression, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089

I will try to create simple repro app on these weekends and will open a new issue.
Thanks for quick response.

**Comment by bthharper** [comment]
I have have been getting the same issue again too, coverlet seems to hang whilst instrumenting the assemblies.  Attaching show it stuck in the MetadataResolver.

![image](https://user-images.githubusercontent.com/8340092/117014994-416dbb80-ace9-11eb-9dca-fdb218db60a8.png)

Is there a debug version of coverlet I can debug with?

**Comment by MarcoRossignoli** [comment]
@bthharper are you on .NET Framework? Are you able to provide a repro? If you're on last version shouldn't be related to reachability helper.

**Comment by bthharper** [comment]
@MarcoRossignoli - thanks for the update I will try and collect some "useful" information, I am running coverlet installed as part of `dotnet tools`

```powershell
PS C:\svn\SunSystemsDotNet\Reporting> dotnet tool list -g
Package Id               Version         Commands
------------------------------------------------------------
coverlet.console         3.0.3           coverlet
dotnet-dump              5.0.160202      dotnet-dump
dotnet-sonarscanner      5.0.4           dotnet-sonarscanner
PS C:\svn\SunSystemsDotNet\Reporting> coverlet --version
Cross platform .NET Core code coverage tool
3.0.0.0
PS C:\svn\SunSystemsDotNet\Reporting> Get-Command Coverlet

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Application     coverlet.exe                                       3.0.3.0    C:\Users\tharper\.dotnet\tools\coverlet.exe
```

I'll take a look at the troubleshooting guide and report back.

**Comment by bthharper** [comment]
@MarcoRossignoli  - sorry for the delay - have just got around to revisiting this.  I am running NUnit test inside coverlet, and the projects are Project SDK targeting net48.

I've cloned, and built the coverlet code and added it to my projects.  However, I use the coverlet command line to run the unit tests, so when I run my command line it does not use the built code, so although I can attach it does not show any symbols.

Do you have a way to debug whilst running under the coverlet tool? 

I tried running using `dotnet test` but this causes other issues,  so thought it best to focus on the approach I am taking.

Any suggestions much appreciated.

**Comment by bthharper** [comment]
Have worked out how to debug my locally built `coverlet.console` tool.  Now just trying to work out where it is stuck.

**Comment by bthharper** [comment]
The code seems to be blocked on this line (line 306 of Instrumenter):

```csharp
module.Write(stream, new WriterParameters { WriteSymbols = true });
```

**Comment by bthharper** [comment]
@MarcoRossignoli - do you have any suggestions.  It looks like the file is locked, but I am unsure why.

**Comment by twenzel** [comment]
@MarcoRossignoli Same problem on my side.
Running SDK styled test project using net472, NUnit3, NUnitTestAdapter 4.0 (was also hanging with 3.x).

After reviewing the dump I made I see the process is hanging on the same location 
`module.Write(stream, new WriterParameters { WriteSymbols = true });`

![image](https://user-images.githubusercontent.com/500376/121951363-21570080-cd5b-11eb-8039-3a42beaacff0.png)

![image](https://user-images.githubusercontent.com/500376/121951233-f2408f00-cd5a-11eb-9e8a-f642ae34e6f1.png)

**Comment by MarcoRossignoli** [comment]
Ok seems the same issue above also without reachability helper. On .NET Framework sometimes Cecil loops forever trying to resolve some types. At the moment there's no workaround if that type is needed. Just for curiosity can you try to downgrade to the second last version of coverlet?

**Comment by twenzel** [comment]
Same result with coverlet.msbuild 3.0.2.
With 3.0.1 I get more output but still hangs:
```
C:\Users\dtw\.nuget\packages\coverlet.msbuild\3.0.1\build\coverlet.msbuild.targets(39,5): warning : [coverlet] Unable to resolve method reference "System.Void System.DirectoryServices.DirectorySearcher::set_Filter(System.String)", assuming calls to will return [C:\xxx\yy.csproj]
C:\Users\dtw\.nuget\packages\coverlet.msbuild\3.0.1\build\coverlet.msbuild.targets(39,5): warning : [coverlet] Unable to resolve method reference "System.Void System.DirectoryServices.DirectorySearcher::set_PageSize(System.Int32)", assuming calls to will return [C:\xxx\yy.csproj]
C:\Users\dtw\.nuget\packages\coverlet.msbuild\3.0.1\build\coverlet.msbuild.targets(39,5): warning : [coverlet] Unable to resolve method reference "System.DirectoryServices.SearchResultCollection System.DirectoryServices.DirectorySearcher::FindAll()", assuming calls to will return [C:\xxx\yy.csproj]
C:\Users\dtw\.nuget\packages\coverlet.msbuild\3.0.1\build\coverlet.msbuild.targets(39,5): warning : [coverlet] Unable to resolve method reference "System.Collections.IEnumerator System.DirectoryServices.SearchResultCollection::GetEnumerator()", assuming calls to will return [C:\xxx\yy.csproj]
C:\Users\dtw\.nuget\packages\coverlet.msbuild\3.0.1\build\coverlet.msbuild.targets(39,5): warning : [coverlet] Unable to resolve method reference "System.DirectoryServices.ResultPropertyCollection System.DirectoryServices.SearchResult::get_Properties()", assuming calls to will return [C:\xxx\yy.csproj]
C:\Users\dtw\.nuget\packages\coverlet.msbuild\3.0.1\build\coverlet.msbuild.targets(39,5): warning : [coverlet] Unable to resolve method reference "System.Boolean System.DirectoryServices.ResultPropertyCollection::Contains(System.String)", assuming calls to will return [C:\xxx\yy.csproj]
C:\Users\dtw\.nuget\packages\coverlet.msbuild\3.0.1\build\coverlet.msbuild.targets(39,5): warning : [coverlet] Unable to resolve method reference "System.DirectoryServices.ResultPropertyValueCollection System.DirectoryServices.ResultPropertyCollection::get_Item(System.String)", assuming calls to will return [C:\xxx\yy.csproj]
C:\Users\dtw\.nuget\packages\coverlet.msbuild\3.0.1\build\coverlet.msbuild.targets(39,5): warning : [coverlet] Unable to resolve method reference "System.Object System.DirectoryServices.ResultPropertyValueCollection::get_Item(System.Int32)", assuming calls to will return [C:\xxx\yy.csproj]
```
With 2.9.0 it also hangs (without any additional output).

**Comment by Bertk** [comment]
The issue might be out-dated because new releases are available.

Please reopen the issue if the bug still exists and can be reproduced with latest preview from [coverlet nightly build](https://dev.azure.com/tonerdo/coverlet/_artifacts/feed/coverlet-nightly).

see [Consume nightly build](https://github.com/coverlet-coverage/coverlet/blob/9295fca754531d6badd0185d0709d41a6eeddd5f/Documentation/ConsumeNightlyBuild.md)
